File: Binomial Heap/findMinBinomialHeap.c
/* File: findMinBinomialHeap.c */

#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int key;
    struct Node *sibling;
} Node;

/* Function to find minimum key in root list */
Node* findMin(Node* head) {
    if (head == NULL)
        return NULL;

    Node* minNode = head;
    Node* curr = head->sibling;

    while (curr != NULL) {
        if (curr->key < minNode->key)
            minNode = curr;

        curr = curr->sibling;
    }

    return minNode;
}

int main() {

    /* Creating a sample root list of binomial heap */
    Node *n1 = (Node*)malloc(sizeof(Node));
    Node *n2 = (Node*)malloc(sizeof(Node));
    Node *n3 = (Node*)malloc(sizeof(Node));

    n1->key = 10;
    n2->key = 5;
    n3->key = 20;

    n1->sibling = n2;
    n2->sibling = n3;
    n3->sibling = NULL;

    Node* min = findMin(n1);

    if(min != NULL)
        printf("Minimum key in Binomial Heap: %d\n", min->key);

    return 0;
}

File: Binomial Heap/FindMinBinomialHeap.java
// File: FindMinBinomialHeap.java

class Node {
    int key;
    Node sibling;

    Node(int key) {
        this.key = key;
        this.sibling = null;
    }
}

public class FindMinBinomialHeap {

    static Node findMin(Node head) {

        if(head == null)
            return null;

        Node minNode = head;
        Node current = head.sibling;

        while(current != null) {

            if(current.key < minNode.key)
                minNode = current;

            current = current.sibling;
        }

        return minNode;
    }

    public static void main(String[] args) {

        Node n1 = new Node(10);
        Node n2 = new Node(5);
        Node n3 = new Node(20);

        n1.sibling = n2;
        n2.sibling = n3;

        Node min = findMin(n1);

        if(min != null)
            System.out.println("Minimum key in Binomial Heap: " + min.key);
    }
}
